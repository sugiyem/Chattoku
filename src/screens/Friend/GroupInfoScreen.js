import React, { useEffect, useState } from "react";
import { Alert } from "react-native";
import {
  fetchGroupMembers,
  fetchPendingGroupMembers
} from "../../services/Friend/FetchGroup";
import { fetchGroupAdminIDs } from "../../services/Friend/FetchGroupAdmin";
import { groupMemberType } from "../../constants/Group";
import { groupMemberSorter } from "../../services/Friend/Sorter";
import { getCurrentUID } from "../../services/Profile/FetchUserInfo";
import { getAdvancedGroupRole } from "../../services/Friend/GroupRole";
import RenderGroupDetail from "../../components/Friend/RenderGroupDetail";
import Loading from "../../components/Miscellaneous/Loading";

const GroupInfoScreen = ({ navigation, route }) => {
  const [members, setMembers] = useState([]);
  const [pendingMembers, setPendingMembers] = useState([]);
  const [adminIDs, setAdminIDs] = useState([]);
  const [isMemberLoading, setIsMemberLoading] = useState(false);
  const [isPendingLoading, setIsPendingLoading] = useState(false);
  const [isAdminLoading, setIsAdminLoading] = useState(false);
  const groupInfo = route.params.groupData;
  const currentID = getCurrentUID();
  const memberIDs = members.map((user) => user.id);
  const pendingMemberIDs = pendingMembers.map((user) => user.id);

  const renderType = adminIDs.includes(currentID)
    ? currentID === groupInfo.owner
      ? groupMemberType.OWNER
      : groupMemberType.ADMIN
    : groupMemberType.MEMBER;

  const detailedMembers = members
    .map((member) => ({
      ...member,
      groupRole: getAdvancedGroupRole(
        member.id,
        groupInfo.owner,
        adminIDs,
        memberIDs,
        pendingMemberIDs
      )
    }))
    .sort(groupMemberSorter);

  console.log(detailedMembers);

  useEffect(() => {
    setIsMemberLoading(true);

    return fetchGroupMembers({
      groupID: groupInfo.id,
      onSuccess: (data) => {
        setMembers(data);
        setIsMemberLoading(false);
      },
      onFailure: (error) => Alert.alert("Error", error.message)
    });
  }, []);

  useEffect(() => {
    setIsPendingLoading(true);

    return fetchPendingGroupMembers({
      groupID: groupInfo.id,
      onSuccess: (data) => {
        setPendingMembers(data);
        setIsPendingLoading(false);
      },
      onFailure: (error) => Alert.alert("Error", error.message)
    });
  }, []);

  useEffect(() => {
    setIsAdminLoading(true);

    return fetchGroupAdminIDs({
      groupID: groupInfo.id,
      onSuccess: (data) => {
        setAdminIDs(data);
        setIsAdminLoading(false);
      }
    });
  }, []);

  return (
    <Loading isLoading={isAdminLoading || isMemberLoading || isPendingLoading}>
      <RenderGroupDetail
        type={renderType}
        groupInfo={groupInfo}
        members={detailedMembers}
        pendingMembers={pendingMembers}
        navigation={navigation}
      />
    </Loading>
  );
};

export default GroupInfoScreen;
