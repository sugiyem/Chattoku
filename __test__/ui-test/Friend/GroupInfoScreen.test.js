import React from "react";
import renderer from "react-test-renderer";
import GroupInfoScreen from "../../../src/screens/Friend/GroupInfoScreen";

const mockNavigation = {
  goBack: jest.fn(),
  navigate: jest.fn(),
  push: jest.fn(),
  replace: jest.fn()
};
const mockRoute = {
  params: {
    groupData: {
      id: "weeb",
      name: "Weebs",
      description: "weebs only",
      img: "weeb-img"
    }
  }
};

jest.mock("../../../src/services/Friend/FetchGroup");
jest.mock("../../../src/services/Friend/FetchGroupAdmin");
jest.mock("../../../src/services/Profile/FetchUserInfo");

describe("Test Group Info Screen UI", () => {
  it("Renders correctly", () => {
    const tree = renderer
      .create(<GroupInfoScreen navigation={mockNavigation} route={mockRoute} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
