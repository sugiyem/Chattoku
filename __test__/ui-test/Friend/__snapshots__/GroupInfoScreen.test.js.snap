// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Group Info Screen UI Renders correctly 1`] = `
<ViewManagerAdapter_ExpoLinearGradient
  proxiedProperties={
    Object {
      "colors": Array [
        4287090411,
        4283256319,
      ],
      "endPoint": Array [
        1,
        1,
      ],
      "locations": undefined,
      "startPoint": Array [
        0,
        0,
      ],
    }
  }
  style={
    Object {
      "flex": 1,
      "padding": 5,
    }
  }
>
  <RCTScrollView
    style={
      Object {
        "flex": 1,
        "padding": 5,
      }
    }
  >
    <View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#7FFFD4",
            "borderRadius": 10,
            "borderWidth": 1,
            "marginBottom": 5,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 5,
            "opacity": 1,
            "paddingBottom": 5,
            "paddingLeft": 5,
            "paddingRight": 5,
            "paddingTop": 5,
          }
        }
        testID="backButton"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#483d8b",
                "fontSize": 14,
                "textAlign": "center",
              },
            ]
          }
        >
          Go Back
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "marginTop": 10,
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "uri": "weeb-img",
            }
          }
          style={
            Array [
              Object {
                "backgroundColor": "#ffffff",
                "borderColor": "#000000",
                "borderRadius": 75,
                "borderWidth": 1,
                "height": 150,
                "marginBottom": 10,
                "width": 150,
              },
            ]
          }
          testID="image"
        />
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "Gill Sans",
                "fontSize": 20,
                "fontWeight": "bold",
                "marginVertical": 10,
                "textAlign": "center",
              },
            ]
          }
          testID="name"
        >
          Weebs
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#0018a8",
                "fontSize": 15,
                "fontWeight": "600",
                "marginBottom": 10,
                "textAlign": "center",
              },
            ]
          }
          testID="description"
        >
          weebs only
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "borderTopColor": "#000000",
              "borderTopWidth": 1,
              "marginBottom": 10,
              "marginLeft": 10,
              "marginRight": 10,
              "marginTop": 10,
              "paddingBottom": 5,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 5,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="accordion"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#87CEFA",
                "borderBottomWidth": 1,
                "borderColor": "#011F5B",
                "borderRadius": 10,
                "borderWidth": 1,
                "flexDirection": "row",
                "marginVertical": 5,
                "padding": 14,
              }
            }
            testID="padView"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                Object {
                  "colors": Object {
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": Object {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 17,
                  }
                }
                testID="listItemTitle"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Gill Sans",
                        "fontSize": 20,
                        "fontWeight": "600",
                        "textDecorationLine": "underline",
                      },
                    ]
                  }
                  testID="title"
                >
                  Members
                </Text>
              </Text>
            </View>
            <View
              style={
                Object {
                  "paddingLeft": 16,
                }
              }
            />
            <View
              collapsable={false}
              style={
                Object {
                  "transform": Array [
                    Object {
                      "rotate": "0deg",
                    },
                  ],
                }
              }
            >
              <View
                style={
                  Object {
                    "overflow": "hidden",
                  }
                }
              >
                <View>
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "maxHeight": "0%",
              "opacity": 0,
            }
          }
        />
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="accordion"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#87CEFA",
                "borderBottomWidth": 1,
                "borderColor": "#011F5B",
                "borderRadius": 10,
                "borderWidth": 1,
                "flexDirection": "row",
                "marginVertical": 5,
                "padding": 14,
              }
            }
            testID="padView"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                Object {
                  "colors": Object {
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": Object {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 17,
                  }
                }
                testID="listItemTitle"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Gill Sans",
                        "fontSize": 20,
                        "fontWeight": "600",
                        "textDecorationLine": "underline",
                      },
                    ]
                  }
                  testID="title"
                >
                  Pending Members
                </Text>
              </Text>
            </View>
            <View
              style={
                Object {
                  "paddingLeft": 16,
                }
              }
            />
            <View
              collapsable={false}
              style={
                Object {
                  "transform": Array [
                    Object {
                      "rotate": "0deg",
                    },
                  ],
                }
              }
            >
              <View
                style={
                  Object {
                    "overflow": "hidden",
                  }
                }
              >
                <View>
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "maxHeight": "0%",
              "opacity": 0,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</ViewManagerAdapter_ExpoLinearGradient>
`;
